### PR TITLE
Extract dynamically defined method content to real methods

### PR DESCRIPTION
**Pull request by Fryguy**
This is done to reduce the complexity score and to find where the real
issues may lie.

Flog scores:

Before:
```
   254.0: flog total
    16.9: flog/method average

   154.1: Kubeclient::Client#none
```

After:
```
   241.3: flog total
    11.5: flog/method average

    65.9: Kubeclient::Client#none
    27.1: Kubeclient::WatchStream#each     lib/kubeclient/watch_stream.rb:11
    23.1: main#none
    20.0: Kubeclient::Client#get_all_for_entity lib/kubeclient.rb:42
    13.6: Kubeclient::Client#update_for_entity lib/kubeclient.rb:96
```

### CODE REVIEW

File: lib/kubeclient.rb [hunk 1]
```diff
@@ -39,6 +39,73 @@ def handling_kube_exception
       raise KubeException.new(e.http_code, JSON.parse(e.response)['message'])
     end
 
+    def get_all_for_entity(entity)
```
Comment (abonas) [inline lib/kubeclient.rb]: method name is confusing imho

**Comment by abonas** [comment]
@Fryguy can you please explain why does it reduce complexity? 
I am definitely supporting simplification of the longer methods, no doubt. But I don't see much difference in the majority of the methods, they were 3-5 lines anyhow, so why is this considered a simplification?

**Comment by Fryguy** [comment]
@abonas complexity is basically defined by the number of code paths.  Generally the higher the complexity,  the less readable and the harder to test.

Most complexity algorithms consider calling a method low complexity.   It doesn't matter how complex the method actually is.   By defining a method,  you are creating a line of separation.   On the other side of the line,  you can test in isolation, and with a good method name make the calling code much more readable. 

So in this case,  by extracting out the guts of the methods,  we are

- isolating the code to a method which can be tested in isolation (and able to be analysed for complexity in isolation)
- separating the metaprogramming part from the "real" code part,  which helps readability
- there is a scope attached to each of those blocks previously, that is wider than is needed by the method.  That is, the blocks have access to variables they don't need from the class scope,  and can even modify them.  By isolating to a method, that scoping is eliminated. 

See also  http://blog.codeclimate.com/blog/2013/08/07/deciphering-ruby-code-metrics/

**Comment by Fryguy** [comment]
Rebased the PR.

File: lib/kubeclient.rb (hunk 1 above)
Comment (Fryguy) [inline lib/kubeclient.rb]: Yeah...do you have a better idea?  Maybe just `get_entities` to match the previous `get_pods`, `get_nodes`, etc.?

**Comment by Fryguy** [comment]
And this is why I don't think rubocop should be part of pass/fail for the tests.

**Comment by Fryguy** [comment]
@abonas Updated the code and renamed the extracted methods as you requested.

**Comment by abonas** [comment]
thanks @Fryguy .
I have 2 suggestions to make this even more clear :)
1. wherever there's a "#{method_name}", it sounds imho more correct to call it "entity_name", because it's already part of defining a method name string and it has a prefix anyway.
2. wherever the newly defined methods have a parameter named entity, I suggest to rename 'entity' to 'entity_type'. because right now for example "get_entities(entity)" doesn't look clear, or "create_entity(entity, entity_config)" also makes wonder why both parameters are needed.
other than that it looks great.

Can you please also explain where/how do you run that flog score from? I probably missed that in between zillion checks and validations that were added recently.

**Comment by abonas** [comment]
@Fryguy just making sure you saw my comments in this PR...

**Comment by Fryguy** [comment]
Oh thanks @abonas.  Somehow I did miss your response.

Those are great suggestions...I'll push an update.  To keep it consistent I might also rename the constant ENTITIES to ENTITY_TYPES, as well as the corresponding variables like `ENTITY_TYPES.each do |entity_type|`.  Is that ok with you?

To run flog...

```
gem install flog
flog lib
```

`flog lib` runs flog against all Ruby files in the lib directory.

**Comment by Fryguy** [comment]
@abonas Updated.  I renamed the ENTITIES constant to ENTITY_TYPES.  If you are not ok with that I'll undo it, but I actually think it's all clearer now.

**Comment by abonas** [comment]
> ENTITIES to ENTITY_TYPES, as well as the corresponding variables like ENTITY_TYPES.each do |entity_type|. Is that ok with you?

sounds great

**Comment by Fryguy** [comment]
Great, then this should be ready to go.

File: lib/kubeclient.rb [hunk 2]
```diff
@@ -17,7 +17,7 @@ module Kubeclient
   # Kubernetes Client
   class Client
     attr_reader :api_endpoint
-    ENTITIES = %w(Pod Service ReplicationController Node Event Endpoint)
+    ENTITY_TYPES = %w(Pod Service ReplicationController Node Event Endpoint)
```
Comment (simon3z) [inline lib/kubeclient.rb]: We recently added Namespaces, is it missing now?

Comment (abonas) [inline lib/kubeclient.rb]: > We recently added Namespaces, is it missing now?

this PR is up to date to yesterday, namespaces were added today. it supposed to rebase with np on top of namespaces PR imho.

**Pull request merged**